### PR TITLE
OD unit test passing turned on with TF

### PR DESCRIPTION
**Pull request by jakesabathia2**
Get rid of MXNET from OD to make it testable.

### CODE REVIEW

File: src/python/turicreate/toolkits/object_detector/object_detector.py [hunk 1]
```diff
@@ -345,7 +352,27 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
...
+            'grid_height': params['grid_shape'][0],
+            'grid_width': params['grid_shape'][1],
+            'max_iterations': num_iterations,
+            'mlmodel_path' : params['mlmodel_path'],
+            'classes' : classes
+
+        }
+        model = _tc.extensions.object_detector()
+        model.train(data=dataset, annotations_column_name=annotations, image_column_name=feature, options=tf_config)
+        return ObjectDetector_beta(model_proxy=model, name="object_detector")
```
Comment (nickjong) [inline src/python/turicreate/toolkits/object_detector/object_detector.py]: Why can't we move this earlier and avoid more of the `if not USE_CPP` cases above?

File: src/python/turicreate/test/test_object_detector.py [hunk 2]
```diff
@@ -199,6 +198,7 @@ def test_create_with_empty_dataset(self):
         with self.assertRaises(_ToolkitError):
             tc.object_detector.create(self.sf[:0])
 
+    @pytest.mark.xfail(IS_PRE_6_0_RC, reason='Requires MXNet')
```
[the remark this answers is not in the document]
Comment (jakesabathia2) [reply src/python/turicreate/test/test_object_detector.py]: Okay I think it doesn't, I just want to mark out those test cases which aren't passes now.

File: src/python/turicreate/toolkits/object_detector/object_detector.py [hunk 3]
```diff
@@ -182,12 +182,17 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
...
-    from ._mx_detector import YOLOLoss as _YOLOLoss
-    from ._model import tiny_darknet as _tiny_darknet
-    from ._sframe_loader import SFrameDetectionIter as _SFrameDetectionIter
-    from ._manual_scheduler import ManualScheduler as _ManualScheduler
-    import mxnet as _mx
-    from .._mxnet import _mxnet_utils
+
+    if not USE_CPP:
+        # Use MxNet
+        print('shit')
```
[the remark this answers is not in the document]
Comment (jakesabathia2) [reply src/python/turicreate/toolkits/object_detector/object_detector.py]: Sorry I will make sure all debug message are deleted next time.

[the remark this answers is not in the document]
Comment (Jarvi-Izana) [reply src/python/turicreate/toolkits/object_detector/object_detector.py]: there are a bunch of profanity filter git plugins to make sure your commit is clean. ;-).

**Comment by jakesabathia2** [comment]
1. Fix batch size issue for OD (pad each data batch to full batch_size).
fix: ObjectDetectorTest::test_dict_annotations, ObjectDetectorTest::test_single_image
2. Fix the crash when an image doesn't have annotations.
fix: ObjectDetectorTest::test_dict_annotations
3. Reset tf graph when create model 
fix: ObjectDetectorTest::test_export_coreml, ObjectDetectorTest::test_export_coreml_with_non_maximum_suppression

**Comment by jakesabathia2** [comment]
Now the OD should run without any crashes and failures.

File: src/ml/neural_net/image_augmentation.cpp [hunk 4]
```diff
@@ -35,6 +35,11 @@ shared_float_array convert_to_shared_float_array(
 
 std::vector<image_annotation> convert_to_image_annotation(
     shared_float_array augmented_annotation) {
+  // Check if the annotation is empty
+  if (augmented_annotation.size() == 1) {
```
Comment (nickjong) [inline src/ml/neural_net/image_augmentation.cpp]: Please check for `.dim()` being 0 instead. Size equal 1 doesn’t read empty in the same way

File: src/ml/neural_net/image_augmentation.cpp [hunk 5]
```diff
@@ -198,8 +204,15 @@ image_augmenter::result float_array_image_augmenter::prepare_images(
   float_array_result augmented_data = prepare_augmented_images(input_to_tf_aug);
 
   // Convert augmented_data to the data structure needed
-  res.image_batch = augmented_data.images;
-  for (size_t a = 0; a < n; a++) {
+  std::vector<float> result_array(n * h * w * c);
+  size_t image_size = augmented_data.images.size();
+  const float* start_address = augmented_data.images.data();
+  const float* end_address = (float*)(start_address + image_size);
```
Comment (nickjong) [inline src/ml/neural_net/image_augmentation.cpp]: The `float *` should be unnecessary 

For future reference, don’t use C-style casts in C++ code anyway

File: src/python/turicreate/toolkits/object_detector/object_detector.py [hunk 6]
```diff
@@ -300,7 +324,6 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
     # to be problematic to run independently of a MXNet-powered neural network
     # in a separate thread. For this reason, we restrict IO threads to when
     # the neural network backend is MPS.
-    io_thread_buffer_size = params['io_thread_buffer_size'] if use_mps else 0
```
Comment (nickjong) [inline src/python/turicreate/toolkits/object_detector/object_detector.py]: Why did this move? (And without its comment?)

File: src/python/turicreate/toolkits/object_detector/object_detector.py [hunk 7]
```diff
@@ -182,12 +182,16 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
...
+
+    if not USE_CPP:
+        # Use MxNet
+        from ._model import tiny_darknet as _tiny_darknet
+        import mxnet as _mx
+        from .._mxnet import _mxnet_utils
+        from ._mx_detector import YOLOLoss as _YOLOLoss
+        from ._sframe_loader import SFrameDetectionIter as _SFrameDetectionIter
+        from ._manual_scheduler import ManualScheduler as _ManualScheduler
+
```
Comment (nickjong) [inline src/python/turicreate/toolkits/object_detector/object_detector.py]: Could we just move these down until they are needed? It seems weird to import them conditionally here

File: src/python/turicreate/toolkits/object_detector/object_detector.py [hunk 8]
```diff
@@ -267,6 +271,28 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
         'mlmodel_path': pretrained_model_path,
     }
 
+    if batch_size < 1:
+        batch_size = 32  # Default if not user-specified
```
Comment (nickjong) [inline src/python/turicreate/toolkits/object_detector/object_detector.py]: The C++ implementation defines the default itself. If the user does not specify a value, pass no batch size to use the heuristic for computing batch size defined in C++

File: src/ml/neural_net/image_augmentation.cpp (hunk 4 above)
Comment (Jarvi-Izana) [reply src/ml/neural_net/image_augmentation.cpp]: it seems like empty annotation is not covered by our python unit test.

Comment (jakesabathia2) [reply src/ml/neural_net/image_augmentation.cpp]: @Jarvi-Izana I think it is covered:
```
for j in range(rs.randint(max_num_boxes_per_image)):
            left, right = np.sort(rs.randint(0, img_shape[1], size=2))
            top, bottom = np.sort(rs.randint(0, img_shape[0], size=2))
```
the bounding box number is generated randomly,
so that why I found this bug ;) .

File: src/ml/neural_net/image_augmentation.cpp (hunk 5 above)
Comment (jakesabathia2) [reply src/ml/neural_net/image_augmentation.cpp]: Good learn ;)

Comment (Jarvi-Izana) [reply src/ml/neural_net/image_augmentation.cpp]: besides, you don't need to cast type for pointer arithmetic.

File: src/ml/neural_net/image_augmentation.cpp (hunk 4 above)
Comment (Jarvi-Izana) [reply src/ml/neural_net/image_augmentation.cpp]: lol, good

File: src/python/turicreate/test/test_object_detector.py [hunk 9]
```diff
@@ -329,6 +328,7 @@ def test_export_coreml(self):
 
     @unittest.skipIf(_mac_ver() < (10, 14),
         "Non-maximum suppression is only supported on MacOS 10.14+.")
+
```
Comment (nickjong) [inline src/python/turicreate/test/test_object_detector.py]: Why are we adding a blank line here? Does it affect the scope of the annotation?

Comment (jakesabathia2) [reply src/python/turicreate/test/test_object_detector.py]: I'll delete it
